File: src/app/services/prediction/PredictionService.ts
import {
  Bet,
  BetPredictionContext,
} from '../../types/prediction/BetPredictionContext';
import {
  BetResultInfluencer,
  PrecheckResult,
} from '../../types/prediction/BetResultInfluencer';
import FootyStatsBttsYesPredictionInfluencer from './influencer/FootyStatsBttsYesPredictionInfluencer';
import FootyStatsOverFTPredictionInfluencer from './influencer/FootyStatsOverFTPredictionInfluencer';
import LeaguePositionDiffInfluencer from './influencer/LeaguePositionDiffInfluencer';
import LeaguePositionInfluencer from './influencer/LeaguePositionInfluencer';
import OddsBttsYesInfluencer from './influencer/OddsBttsYesInfluencer';
import OddsGoalsOverInfluencer from './influencer/OddsGoalsOverInfluencer';
import {
  InfluencerResult,
  PredictionAnalyze,
  PredictionResult,
} from './PredictionService.types';

/**
 * All known prediction influencer
 */
const betResultInfluencer: BetResultInfluencer[] = [
  new LeaguePositionInfluencer(),
  new LeaguePositionDiffInfluencer(),
  new OddsGoalsOverInfluencer(),
  new FootyStatsOverFTPredictionInfluencer(),
  new OddsBttsYesInfluencer(),
  new FootyStatsBttsYesPredictionInfluencer(),
];

function analyze(
  ctx: BetPredictionContext,
  didPredictionCalculation: boolean,
  betOnThis: boolean
): PredictionAnalyze {
  if (didPredictionCalculation === false) {
    return 'NOT_PREDICTED';
  }

  if (ctx.match['Match Status'] !== 'complete') {
    return 'NOT_COMPLETED';
  }

  switch (ctx.bet) {
    case Bet.OVER_ZERO_FIVE: {
      const goals =
        ctx.match['Result - Away Team Goals'] +
        ctx.match['Result - Home Team Goals'];
      if (goals > 0 && betOnThis) {
        return 'SUCCESS';
      }

      return 'FAILED';
    }
    case Bet.BTTS_YES: {
      if (
        (ctx.match['Result - Away Team Goals'] > 0 &&
          ctx.match['Result - Home Team Goals'] > 0 &&
          betOnThis) ||
        ((ctx.match['Result - Away Team Goals'] === 0 ||
          ctx.match['Result - Home Team Goals'] === 0) &&
          !betOnThis)
      ) {
        return 'SUCCESS';
      }

      return 'FAILED';
    }
    default:
      return 'NOT_ANALYZED';
  }
}

export default function prediction(
  ctx: BetPredictionContext
): PredictionResult {
  let result = 0;

  const influencerDetailedResult: InfluencerResult[] = [];

  let doneInfluencerCalculations = 0;
  betResultInfluencer.forEach((influencer) => {
    const preCheckResult = influencer.preCheck(ctx);
    if (preCheckResult === PrecheckResult.OK) {
      const predictionInfluence = influencer.calculateInfluence(ctx);
      result += predictionInfluence.amount;
      doneInfluencerCalculations += 1;
      influencerDetailedResult.push({
        precheckResult: preCheckResult,
        influencerPredictionValue: predictionInfluence.amount,
        influencerName: influencer.influencerName(),
      });
    } else if (
      preCheckResult === PrecheckResult.NOT_ENOUGH_INFORMATION ||
      preCheckResult === PrecheckResult.EXCEPTION
    ) {
      influencerDetailedResult.push({
        influencerName: influencer.influencerName(),
        precheckResult: preCheckResult,
      });
    }
  });

  if (doneInfluencerCalculations > 0) {
    result = Math.round(result / doneInfluencerCalculations);
  }
  const betOnThis = result > 50;
  return {
    betSuccessInPercent: result,
    betOnThis,
    analyzeResult: analyze(ctx, doneInfluencerCalculations > 0, betOnThis),
    influencerDetailedResult,
  };
}

File: footystats-tools-backend/src/main/java/de/ludwig/footystats/tools/backend/services/prediction/PredictionService.java
package de.ludwig.footystats.tools.backend.services.prediction;

import de.ludwig.footystats.tools.backend.services.stats.MatchStatus;
import de.ludwig.footystats.tools.backend.services.prediction.influencer.*;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class PredictionService {

	private final BetResultInfluencer[] betResultInfluencer = {
			new OddsBttsYesInfluencer(),
			new OddsGoalOverInfluencer(),
			new FootyStatsBttsYesPredictionInfluencer(),
			new FootyStatsOverFTPredictionInfluencer()
	};

	public PredictionAnalyze analyze(
			BetPredictionContext ctx,
			boolean didPredictionCalculation,
			boolean betOnThis) {
		if (didPredictionCalculation == false) {
			return PredictionAnalyze.NOT_PREDICTED;
		}

		if (ctx.match().getMatchStatus() != MatchStatus.complete) {
			return PredictionAnalyze.NOT_COMPLETED;
		}

		switch (ctx.bet()) {
			case OVER_ZERO_FIVE: {
				var goals = ctx.match().getResultAwayTeamGoals() +
						ctx.match().getResultHomeTeamGoals();
				if (goals > 0 && betOnThis) {
					return PredictionAnalyze.SUCCESS;
				}

				return PredictionAnalyze.FAILED;
			}
			case BTTS_YES: {
				if ((ctx.match().getResultAwayTeamGoals() > 0 &&
						ctx.match().getResultHomeTeamGoals() > 0 &&
						betOnThis) ||
						((ctx.match().getResultAwayTeamGoals() == 0 ||
								ctx.match().getResultHomeTeamGoals() == 0) &&
								!betOnThis)) {
					return PredictionAnalyze.SUCCESS;
				}

				return PredictionAnalyze.FAILED;
			}
			default:
				return PredictionAnalyze.NOT_ANALYZED;
		}
	}

	public PredictionResult prediction(BetPredictionContext ctx) {
		var result = 0F;

		List<InfluencerResult> influencerDetailedResult = new ArrayList<>();

		var doneInfluencerCalculations = 0;

		for (BetResultInfluencer influencer : betResultInfluencer) {

			var preCheckResult = influencer.preCheck(ctx);
			if (preCheckResult == PrecheckResult.OK) {
				var predictionInfluence = influencer.calculateInfluence(ctx);
				result += predictionInfluence;
				doneInfluencerCalculations += 1;
				influencerDetailedResult
						.add(new InfluencerResult(influencer.influencerName(), predictionInfluence, preCheckResult));
			} else if (preCheckResult == PrecheckResult.NOT_ENOUGH_INFORMATION ||
					preCheckResult == PrecheckResult.EXCEPTION) {
				influencerDetailedResult.add(new InfluencerResult(influencer.influencerName(), 0F, preCheckResult));
			}
		}
		;

		if (doneInfluencerCalculations > 0) {
			result = Math.round(result / doneInfluencerCalculations);
		}
		var betOnThis = result > 50;
		return new PredictionResult(result, betOnThis, analyze(ctx, doneInfluencerCalculations > 0, betOnThis),
				influencerDetailedResult);
	}
}
